components/brave_wallet/browser/solana_data_decoder_utils.cc

```diff
@@ -0,0 +1,204 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/browser/solana_data_decoder_utils.h"
+
+#include <tuple>
+#include <utility>
+
+#include "base/containers/flat_map.h"
+#include "base/containers/span.h"
+#include "base/no_destructor.h"
+#include "base/notreached.h"
+#include "base/sys_byteorder.h"
+#include "brave/components/brave_wallet/common/brave_wallet_constants.h"
+#include "brave/components/brave_wallet/common/solana_utils.h"
+#include "build/build_config.h"
+#include "components/grit/brave_components_strings.h"
+#include "ui/base/l10n/l10n_util.h"
+
+namespace brave_wallet {
+
+constexpr uint8_t kAuthorityTypeMax = 3;
+constexpr size_t kMaxStringSize32Bit = 4294967291u;
+
+absl::optional<uint8_t> DecodeUint8(const std::vector<uint8_t>& input,
+                                    size_t& offset) {
+  if (offset >= input.size() || input.size() - offset < sizeof(uint8_t)) {
```

Comment:
Can we run the functions in this file with this length check inside of a separate thread for parsing so that we can adhere with the rule of 2? Given we're having to manually check lengths here this seems like something arbitrary enough that the safest way for us to do this in C++ is to sandbox it. Otherwise we could utilize a rust library for this, but that's probably more work and comes with the tradeoff of having to add and vet another external dependency.